services/feedz/feedz.service.js

```diff
@@ -0,0 +1,109 @@
+'use strict'
+
+const Joi = require('joi')
+const RouteBuilder = require('../route-builder')
+const { BaseJsonService, NotFound } = require('..')
+const {
+  renderVersionBadge,
+  searchServiceUrl,
+  stripBuildMetadata,
+  selectVersion,
+} = require('../nuget/nuget-helpers')
+
+const schema = Joi.object({
+  items: Joi.array()
+    .items(
+      Joi.object({
+        items: Joi.array().items(
+          Joi.object({
+            catalogEntry: Joi.object({
+              version: Joi.string().required(),
+            }).required(),
+          })
+        ),
+      }).required()
+    )
+    .max(1)
+    .default([]),
+}).required()
+
+function apiUrl({ organization, repository }) {
+  return `https://f.feedz.io/${organization}/${repository}/nuget`
+}
+
+class FeedzVersionService extends BaseJsonService {
+  static category = 'version'
+
+  static route = new RouteBuilder({ base: 'feedz' })
+    .push('(v|vpre)', 'which')
+    .push('([^/]+)', 'organization')
+    .push('([^/]+)', 'repository')
+    .push('(.+?)', 'packageName')
+    .toObject()
+
+  static examples = [
+    {
+      title: 'Feedz',
+      pattern: 'v/:organization/:repository/:packageName',
+      namedParams: {
+        organization: 'shieldstests',
+        repository: 'mongodb',
+        packageName: 'MongoDB.Driver.Core',
+      },
+      staticPreview: this.render({ version: '2.10.4' }),
+    },
+    {
+      title: 'Feedz (with prereleases)',
+      pattern: 'vpre/:organization/:repository/:packageName',
+      namedParams: {
+        organization: 'shieldstests',
+        repository: 'mongodb',
+        packageName: 'MongoDB.Driver.Core',
+      },
+      staticPreview: this.render({ version: '2.11.0-beta2' }),
+    },
+  ]
+
+  static defaultBadgeData = {
+    label: 'feedz',
+  }
+
+  static render(props) {
+    return renderVersionBadge(props)
+  }
+
+  async fetch({ baseUrl, packageName }) {
+    const registrationsBaseUrl = await searchServiceUrl(
+      baseUrl,
+      'RegistrationsBaseUrl'
+    )
+    const json = await this._requestJson({
+      schema,
+      url: `${registrationsBaseUrl}${packageName}/index.json`,
+    })
+
+    if (json.items.length === 1) {
+      return json.items[0].items.map(i =>
+        stripBuildMetadata(i.catalogEntry.version)
+      )
+    } else {
+      throw new NotFound({ prettyMessage: 'not found' })
+    }
```

Comment:
I know this is similar to how the nuget-v3 family was structured, but this is a case where I think we should take advantage of some of the comparatively newer patterns we've been leveraging in our service classes.

For better/cleaner separation of concerns and encapsulation we prefer to use a separate `transform` function that takes the raw response from the upstream provider and does whatever processing/extracting/transforming is needed to get the desired data point (which would be the target `version` in this case).

An added benefit of this approach is that it makes testing logic much easier with standard unit tests, as a `transform` function can be hit directly without having to deal with cumbersome mocking

Reply:
I totally agree. Haven't seen that in most of the services that I looked but the testability is really beneficial. Fixed (and added the tests).